src/algebra/group/freiman.lean

```diff
@@ -0,0 +1,375 @@
+    by rw [multiset.map_const, multiset.map_const, prod_repeat, prod_repeat, hs, ht] }
+
+@[simp, to_additive] lemma const_apply (n : ℕ) (b : β) (x : α) : const A n b x = b := rfl
+
+@[simp, to_additive]
+lemma const_comp (n : ℕ) (c : γ) (f : A →*[n] β) {hf} : (const B n c).comp f hf = const A n c := rfl
+
+/-- `1` is the Freiman homomorphism sending everything to `1`. -/
+@[to_additive "`0` is the Freiman homomorphism sending everything to `0`."]
+instance : has_one (A →*[n] β) := ⟨const A n 1⟩
+
+@[simp, to_additive] lemma one_apply (x : α) : (1 : A →*[n] β) x = 1 := rfl
+
+@[simp, to_additive] lemma one_comp (f : A →*[n] β) {hf} : (1 : B →*[n] γ).comp f hf = 1 := rfl
+
+@[to_additive] instance : inhabited (A →*[n] β) := ⟨1⟩
+
+/-- `f * g` is the Freiman homomorphism  sends `x` to `f x * g x`. -/
+@[to_additive "`f + g` is the Freiman homomorphism sending `x` to `f x + g x`."]
+instance : has_mul (A →*[n] β) :=
+⟨λ f g, { to_fun := λ x, f x * g x,
+  map_prod_eq_map_prod' := λ s t hsA htA hs ht h,
+    by rw [prod_map_mul, prod_map_mul, map_prod_eq_map_prod f hsA htA hs ht h,
+           map_prod_eq_map_prod g hsA htA hs ht h] }⟩
+
+@[simp, to_additive] lemma mul_apply (f g : A →*[n] β) (x : α) : (f * g) x = f x * g x := rfl
+
+@[to_additive] lemma mul_comp (g₁ g₂ : B →*[n] γ) (f : A →*[n] β) {hg hg₁ hg₂} :
+  (g₁ * g₂).comp f hg = g₁.comp f hg₁ * g₂.comp f hg₂ := rfl
+
+/-- If `f` is a Freiman homomorphism to a commutative group, then `f⁻¹` is the Freiman homomorphism
+sending `x` to `(f x)⁻¹`. -/
+@[to_additive]
+instance : has_inv (A →*[n] G) :=
+⟨λ f, { to_fun := λ x, (f x)⁻¹,
+  map_prod_eq_map_prod' := λ s t hsA htA hs ht h,
+    by rw [prod_map_inv', prod_map_inv', map_prod_eq_map_prod f hsA htA hs ht h] }⟩
+
+@[simp, to_additive] lemma inv_apply (f : A →*[n] G) (x : α) : f⁻¹ x = (f x)⁻¹ := rfl
+
+@[simp, to_additive] lemma inv_comp (f : B →*[n] G) (g : A →*[n] β) {hf hf'} :
+  f⁻¹.comp g hf = (f.comp g hf')⁻¹ :=
+ext $ λ x, rfl
+
+/-- If `f` and `g` are Freiman homomorphisms to a commutative group, then `f / g` is the Freiman
+homomorphism sending `x` to `f x / g x`. -/
+@[to_additive "If `f` and `g` are additive Freiman homomorphisms to an additive commutative group,
+then `f - g` is the additive Freiman homomorphism sending `x` to `f x - g x`"]
+instance : has_div (A →*[n] G) :=
+⟨λ f g, { to_fun := λ x, f x / g x,
+  map_prod_eq_map_prod' := λ s t hsA htA hs ht h,
+    by rw [prod_map_div, prod_map_div, map_prod_eq_map_prod f hsA htA hs ht h,
+           map_prod_eq_map_prod g hsA htA hs ht h] }⟩
+
+@[simp, to_additive] lemma div_apply (f g : A →*[n] G) (x : α) : (f / g) x = f x / g x := rfl
+
+@[simp, to_additive] lemma div_comp (f₁ f₂ : B →*[n] G) (g : A →*[n] β) {hf hf₁ hf₂} :
+  (f₁ / f₂).comp g hf = f₁.comp g hf₁ / f₂.comp g hf₂ :=
+ext $ λ x, rfl
+
+/-! ### Instances -/
+
+/-- `A →*[n] β` is a `comm_monoid`. -/
+@[to_additive "`α →+[n] β` is an `add_comm_monoid`."]
+instance : comm_monoid (A →*[n] β) :=
```

Comment:
Is it a pullback from the structure on the space of functions? If yes, then you can use `injective.comm_monoid` (you might need to add a version with `pow` suffix).